File: app/src/main/java/com/nikola/jakshic/dagger/viewModel/MatchViewModel.java
package com.nikola.jakshic.dagger.viewModel;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.arch.paging.PagedList;

import com.nikola.jakshic.dagger.Status;
import com.nikola.jakshic.dagger.model.match.Match;
import com.nikola.jakshic.dagger.repository.MatchRepository;

import javax.inject.Inject;

public class MatchViewModel extends ViewModel {

    private LiveData<PagedList<Match>> list;
    private MatchRepository repository;
    private MutableLiveData<Status> status;
    private boolean initialFetch;

    @Inject
    public MatchViewModel(MatchRepository repository) {
        this.repository = repository;
        status = new MutableLiveData<>();
        status.setValue(Status.LOADING);
    }

    public void initialFetch(long id) {
        if (!initialFetch) {
            list = repository.getMatches(status,id);
            repository.refreshMatches(status, id);
            initialFetch = true;
        }
    }

    public void fetchHeroes(long id) {
        repository.refreshMatches(status, id);
    }

    public LiveData<PagedList<Match>> getMatches() {
        return list;
    }

    public MutableLiveData<Status> getStatus() {
        return status;
    }
}

File: app/src/main/java/com/nikola/jakshic/dagger/ui/profile/matches/MatchViewModel.kt
package com.nikola.jakshic.dagger.ui.profile.matches

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.paging.PagedList
import com.nikola.jakshic.dagger.repository.MatchRepository
import com.nikola.jakshic.dagger.ui.ScopedViewModel
import com.nikola.jakshic.dagger.ui.Status
import com.nikola.jakshic.dagger.vo.Match
import kotlinx.coroutines.experimental.launch
import javax.inject.Inject

class MatchViewModel @Inject constructor(private val repository: MatchRepository) : ScopedViewModel() {

    private lateinit var response: Response

    val list: LiveData<PagedList<Match>>
        get() = response.pagedList

    private val _refreshStatus = MutableLiveData<Status>()
    val refreshStatus: LiveData<Status>
        get() = _refreshStatus

    val loadMoreStatus: LiveData<Status>
        get() = response.status

    private var initialFetch = false

    private val onSuccess: () -> Unit = { _refreshStatus.value = Status.SUCCESS }
    private val onError: () -> Unit = { _refreshStatus.value = Status.ERROR }

    fun initialFetch(id: Long) {
        if (!initialFetch) {
            initialFetch = true
            response = repository.getMatchesLiveData(this, id)
            fetchMatches(id)
        }
    }

    fun fetchMatches(id: Long) {
        launch {
            _refreshStatus.value = Status.LOADING
            repository.fetchMatches(id, onSuccess, onError)
        }
    }

    fun retry() = response.retry()
}